### ISSUE TITLE
Enhance bylaw logic for Voting app to allow more customized quorum and approval rules

### ISSUE THREAD

**Issue by stellarmagnet** [aragon/aragon-apps#49]
## Problem
Based on the Voting specification: http://wiki.aragon.one/dev/apps/voting/

It seems that DAOs will be forced to measure "quorum" based on the number of acceptance votes, as opposed to more standard logic for what "quorum" actually means. 

Whenever you use "Quorum" terminology in the Aragon app, it will be very confusing if you are treating Quorum to mean "minimum acceptance quorum", as this does not match up to the definition of quorum for organizations/businesses, which is:

"Fixed minimum number of eligible members or stockholders (shareholders) who must be present (physically or by proxy) at a meeting before any official business may be transacted or a decision taken therein becomes legally binding." Read more here: http://www.businessdictionary.com/definition/quorum.html

## Solution
In addition to this problem with the terminology, this quorum/acceptance logic should be more flexible. I see three use cases for how a DAO can formulate metrics for approving proposals:

### Disable Quorum
If a DAO does not care to consider "Quorum" as a metric, then it can be disabled in the bylaws, and Approval percentage based on total token supply can be used. Quorum terminology should never be used in DAOs that do not care about true Quorum requirements. [this is equivalent to minimum acceptance quorum]

### Enable Quorum
Whenever Quorum is enabled, this means the Approval Percentage is based on the all of the tokens that voted, and not on the total token supply.

### Modified Quorum
If a DAO wants to be more flexible with how proposals are passed, then allow a feature called "Modified Quorum" which will serve as a conditional logic to determine when the Quorum metric should be considered and how it is treated in relation to Approval Percentage. For example, the DAO can define two ways for proposals to pass. After the time period has ended, if either condition is met, the proposal is approved:
- if Quorum percentage is met & Approval Percentage (based on quorum) is met
- if Yes votes = Quorum Percentage * Approval Percentage [this is equivalent to the "minimum acceptance quorum" variable used in the current Aragon logic]

Here is an example of how you can accommodate for the third use case. When I have more time I can try to simplify the language & mockup a screen that would accommodate to all 3 use cases I described above, but if you shared your current Bylaw screens (if you have any), that'd be helpful.

---
![image](https://user-images.githubusercontent.com/2584493/34556538-8c5961f2-f0eb-11e7-9358-1f053e80d70c.png)

**Comment by lyricalpolymath** [comment]
I like the direction and I would take it a little bit further:
It appears clear to me that there are **many kinds of possible "quorum"** (also considering the proposal for an [Exponential Backoff Quorum Discovery](https://github.com/aragon/governance/issues/17) mechanism

a possible solution to this would be to **add a descriptor to the word "quorum"** such as:
- Share Quorum or Token Quorum
- Participating Member Quorum
- Discover Quorum or Quorum Discovery
- No Quorum

These are mutually exclusive options (ie radio buttons)

It's clear that in different occasions One would want to make decisions based on the token holders and another more inclusive based on the actual members of the community participating

The first 2 (Token Quorum and Participating Member Quorum), when selected, would make another field appear : the "minimum quorum" that the user wants for that voting

an advantage of this setup is that
it's possible that **quorum mechanisms can be developed by developers and hooked, as add-ons / modules, into the interface to add more options to the list of possible quorums**
it would allow a lot of experimentation
(at a high level I think it would be enough to decouple the actual voting from the voting counting so that it's the voting counting smart contract that actually contains the quorum mechanism)

**Comment by sohkai** [comment]
@stellarmagnet I think some of the confusion arises because we used "quorum" in a [distributed-systems manner](https://en.wikipedia.org/wiki/Quorum_(distributed_computing)) rather than in a business context (of course we did, as engineers 😉). But clarity here is really important and the Dapp client should make this as obvious as possible while offering as much flexibility as it can to users.

@lyricalpolymath 

> (Participating Member Quorum): based on the actual members of the community participating

This is pretty difficult to get right as it's a sybil attack vector.


----------------------

Something that we could do in the future to improve flexibility is allow each vote or the entire voting app to outsource its approval and quorum calculations to another contract that users could plug in themselves (and we'd provide a few of the more common use cases as part of the app).

The pieces of information useful to any such contract would include:
- number of total token votes
- number of token approvals
- total token supply
- number of total voters (again, difficult right now to ensure sybil resistance)
- number of voters approving
